### PR TITLE
fix(40322): bad completions for string literal type as indexed access type object

### LINKED ISSUE #40322: Bad completions for string literal type as indexed access type object
Found while reading over #39697

```ts
let x: ""["bar"]
```

Request completions inside of `""` and you'll get the names of every propery on `String`. What's happening is that we're not checking to see if the current string literal type is the argument type (as opposed to the object type itself).

Learning how this works might be a bit involved, but the fix is easy: add fourslash tests, add a single `if` check at https://github.com/microsoft/TypeScript/blob/d75894d85b8fd5f83116d3c02b5ba30a62af4348/src/services/stringCompletions.ts#L117-L124

### PR DESCRIPTION
**Pull request by Vboivin**
Fixes #40322 

`let x: "/*foo*/"["bar"];`

Added validation to prevent completion on a string literal when it is the object type.
The cursor at the marker foo will not have autocompletion.

* [x] There is an associated issue in the `Backlog` milestone (**required**)
* [x] Code is up-to-date with the `master` branch
* [x] You've successfully run `gulp runtests` locally
* [x] There are new or updated unit tests validating the change

### CODE REVIEW

**Comment by Vboivin** [comment]
@DanielRosenwasser here is the PR for #40322

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]
1. Use spaces after `if`s
2. You shouldn't need a type assertion to `any` anywhere.
3. Just use an early bail-out to return `undefined` when you're in the wrong node.

File: src/services/stringCompletions.ts [hunk 1]
```diff
@@ -130,7 +130,15 @@ namespace ts.Completions.StringCompletions {
                         //          bar: string;
                         //      }
                         //      let x: Foo["/*completion position*/"]
-                        return stringLiteralCompletionsFromProperties(typeChecker.getTypeFromTypeNode((grandParent as IndexedAccessTypeNode).objectType));
+                        const { objectType } = grandParent as IndexedAccessTypeNode;
+                        let isNodeAnObjectTypeStringLiteral;
+                        if((objectType as any).literal){
+                            isNodeAnObjectTypeStringLiteral = node.text === (objectType as any).literal.text;
```
Comment (DanielRosenwasser) [inline src/services/stringCompletions.ts:136]: Don't check the text, check if `parent !== grandParent.indexType` and just return `undefined`.

Comment (Vboivin) [reply src/services/stringCompletions.ts]: @DanielRosenwasser I applied your suggestion and I also added a check for parenthesized types.

File: src/services/stringCompletions.ts [hunk 2]
```diff
@@ -130,7 +130,11 @@ namespace ts.Completions.StringCompletions {
                         //          bar: string;
                         //      }
                         //      let x: Foo["/*completion position*/"]
-                        return stringLiteralCompletionsFromProperties(typeChecker.getTypeFromTypeNode((grandParent as IndexedAccessTypeNode).objectType));
+                        const { indexType, objectType } = grandParent as IndexedAccessTypeNode;
+                        if (!isParenthesizedTypeNode(indexType) && parent !== indexType) {
```
[the remark this answers is not in the document]
Comment (DanielRosenwasser) [reply src/services/stringCompletions.ts:134]: Or `!rangeContainsPosition(indexType, position)`

File: tests/cases/fourslash/completionListsStringLiteralTypeAsIndexedAccessTypeObject.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+});
+
+const [
+  { fileName },
+  _,
+  __,
+  ___,
+  fifthCaseMarker,
+  sixthCaseMarker,
+] = test.markers();
```
Comment (DanielRosenwasser) [inline tests/cases/fourslash/completionListsStringLiteralTypeAsIndexedAccessTypeObject.ts:28]:
```suggestion
const test5 = test.marker("case_5");
const test6 = test.marker("case_6");
```

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]

**Review by DanielRosenwasser — APPROVED** [review]

**Pull request merged**